### ISSUE TITLE
[Question] MQTTS and "unknown_ca" error

### ISSUE THREAD

**Issue by andreas-jankowski** [thingsboard/thingsboard#5980]
Hello,

I'm struggeling to get an IoT-Device to connect to TB by MQTTS with PEM cert. I have many devices running fine with the same configuration, but this one stopped two day ago to send data. Since then, I can't get it to connect again. Nothing was changed in software. The only hint I have is from the thingsboard.log when I start the script on the mentioned gateway.

Hope someone can help...

**Component**
 * Generic

**Description**

2022-01-28 15:06:36,642 [nioEventLoopGroup-6-9] ERROR o.t.s.t.mqtt.MqttTransportHandler - [6c0e6b1e-e96d-4e37-beb1-65f17ab14f3e] Unexpected Exception
io.netty.handler.codec.DecoderException: javax.net.ssl.SSLHandshakeException: Received fatal alert: unknown_ca
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:477)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:276)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:722)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:658)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:584)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:496)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: javax.net.ssl.SSLHandshakeException: Received fatal alert: unknown_ca
	at java.base/sun.security.ssl.Alert.createSSLException(Alert.java:131)
	at java.base/sun.security.ssl.Alert.createSSLException(Alert.java:117)
	at java.base/sun.security.ssl.TransportContext.fatal(TransportContext.java:336)
	at java.base/sun.security.ssl.Alert$AlertConsumer.consume(Alert.java:293)
	at java.base/sun.security.ssl.TransportContext.dispatch(TransportContext.java:185)
	at java.base/sun.security.ssl.SSLTransport.decode(SSLTransport.java:172)
	at java.base/sun.security.ssl.SSLEngineImpl.decode(SSLEngineImpl.java:681)
	at java.base/sun.security.ssl.SSLEngineImpl.readRecord(SSLEngineImpl.java:636)
	at java.base/sun.security.ssl.SSLEngineImpl.unwrap(SSLEngineImpl.java:454)
	at java.base/sun.security.ssl.SSLEngineImpl.unwrap(SSLEngineImpl.java:433)
	at java.base/javax.net.ssl.SSLEngine.unwrap(SSLEngine.java:637)
	at io.netty.handler.ssl.SslHandler$SslEngineType$3.unwrap(SslHandler.java:295)
	at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1341)
	at io.netty.handler.ssl.SslHandler.decodeJdkCompatible(SslHandler.java:1234)
	at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1283)
	at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:507)
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:446)
	... 17 common frames omitted

**Environment**
 * OS:  Ubuntu 20.04.3 and version
 * ThingsBoard: 3.3.3PE

**Comment by iwas108** [comment]
Any update on this? I Got this error after upgrading to 3.6.0

**Comment by iwas108** [comment]
My problem was solved. Just come back to share my solution to other unfortunate soul.

Firstly, make sure your self signed certificate is in a good condition. In my case my certif is fine, I follow  the step from good CA, good intermediate CA, and finally the good server certificate. Takes note that openssl have a configuration file that can be selected to create server cert, client cert, email cert, etc. The common name, the domain/DNS and the IP on the SNI section must match. The last one, make sure you put your .crt.pem and .key.pem in /etc/thingsboard/conf/ and set the permission as the same as thingsboard.conf.

Secondly, the root of my problem so I lost the entire night to figure it out, it was the machine that I use to compile my code. Before, I use Ubuntu, everything works fine. Then, I switch to Windows, then everything was puffedup.  I figured it out that my Windows machine results a mess in my CA Root code that I flash to the ESP32. I still don't know what it was, the line ending or what. Just tried to change the line ending in Visual Studio Code to LR and other, but no luck. So the last resort I switch back to Ubuntu and my life is back. I hope this help others.
